otherlibs/unix/channels_win32.c

```diff
@@ -60,15 +60,36 @@ static DWORD check_stream_semantics(value handle)
   }
 }
 
+int caml_win32_get_CRT_fd(value handle)
+{
+  int fd;
+  SPIN_WAIT {
+    fd = atomic_load(&CRT_fd_val(handle));
+    if (fd != GETTING_CRT_FD) return fd;
+  }
+}
+
 int caml_win32_CRT_fd_of_filedescr(value handle)
 {
-  if (CRT_fd_val(handle) != NO_CRT_FD) {
-    return CRT_fd_val(handle);
-  } else {
-    int fd = _open_osfhandle((intptr_t) Handle_val(handle), O_BINARY);
-    if (fd == -1) caml_uerror("channel_of_descr", Nothing);
-    CRT_fd_val(handle) = fd;
-    return fd;
+  SPIN_WAIT {
+    int fd = atomic_load(&CRT_fd_val(handle));
+    switch (fd) {
```

Comment:
There's nothing performance critical about this, but it might be nicer to have the "fast" path first?

```c
if (fd > GETTING_CRT_FD) {
  return fd;
} else if (fd == NO_CRT_FD) {
  ...
}
```
?

Reply:
I find the switch clearer.